File: ruby/test/dialect_test.rb
# frozen_string_literal: true

# Where Ruby's regex dialect differs from Python's, pinned rather than reasoned about.
#
# The detector was written against Python `re` and ported through TypeScript.
# Ruby's defaults disagree with Python's in two ways, each a silent behaviour
# change rather than an error:
#
# * `^` and `$` mean start- and end-of-**line** in Ruby, and start- and
#   end-of-**string** in Python without `re.MULTILINE`. (`\Z` is Python's `$`;
#   `\z` is JavaScript's.)
# * `\w`, `\d` and `\s` are ASCII-only in Ruby and Unicode-aware in Python.
#
# `\b` is the one that does NOT disagree, which is worth a test of its own: the
# TypeScript port spells its boundaries out because *JavaScript's* `\b` is
# ASCII-only, and inheriting that reasoning without checking it would state
# something about Ruby that is not true.
#
# Every site is already written the Python way. This file exists because
# **neither shared spec layer catches it if somebody writes it back.** Measured,
# not assumed: mutating `\z` to `$` in `RELATION_ATTACHED_BEFORE` and `\Z` to `$`
# in `ZIP` leaves all 36 conformance frames green and all 2,526 primitive
# assertions green. The fixture corpus is single-line, and these two rules only
# diverge across a newline.
#
# So each test below asserts the divergence in both directions: that the pattern
# as written gives the reference answer, AND that the idiomatic-Ruby spelling
# gives a different one. The second half is what makes this a test rather than a
# restatement — an assertion that only pins the current answer would still pass
# if the difference evaporated, and then it would be guarding nothing.

require "minitest/autorun"
require "set"

require "vicary"

class DialectTest < Minitest::Test
  C = Vicary::Candidates
  S = Vicary::Structured

  # The same pattern with a Ruby-idiomatic anchor substituted back in.
  def loosened(pattern, from, to)
    source = pattern.source
    refute_equal source, source.sub(from, to),
                 "#{from} is no longer in this pattern, so this test is checking nothing"
    Regexp.new(source.sub(from, to), pattern.options)
  end

  def test_zip_anchors_on_the_string_not_the_line
    # `\s*$` would satisfy the ZIP lookahead at the end of every line of a
    # hard-wrapped essay, so any 5-digit number a student ends a line on — a
    # locker combination, a population, a year range — masks as `{ZIP_CODE_1}`.
    text = "I live at 12345\nMy locker combination is 90210 and I forget it."

    assert_empty text.scan(S::ZIP),
                 "a 5-digit number at a line end is not a ZIP code"
    assert_equal ["12345"], text.scan(loosened(S::ZIP, '\Z', "$")),
                 "the line-anchored spelling was expected to over-fire here; if it " \
                 "no longer does, this guard has stopped guarding anything"

    # ...and the cases a real ZIP arrives in still match, so the anchor is not
    # simply switched off.
    assert_equal ["12345"], "Send it to 12345.".scan(S::ZIP)
    assert_equal ["12345"], "12345".scan(S::ZIP)
    assert_equal ["12345"], "Akron 12345 OH".scan(S::ZIP)
  end

  def test_an_attached_relation_may_not_reach_across_a_line_break
    # The window handed to this pattern is the 90 characters before a name. With
    # `$` the relation cue need only end *a line* rather than run up to the name,
    # so "my cousin" two lines earlier attaches to a name it has nothing to do
    # with — and that wrongly overrides a title keep, or wrongly refuses
    # corroboration for a public figure's surname.
    across_a_break = "my cousin \nand then the dog ran "

    refute C::RELATION_ATTACHED_BEFORE.match?(across_a_break),
           "a relation cue on an earlier line is not attached to this name"
    assert loosened(C::RELATION_ATTACHED_BEFORE, '\z', "$").match?(across_a_break),
           "the line-anchored spelling was expected to attach across the break"

    # Genuinely adjacent still attaches, on both the possessive and the modifier
    # forms the reference accepts.
    assert C::RELATION_ATTACHED_BEFORE.match?("we walked home. my neighbor ")
    assert C::RELATION_ATTACHED_BEFORE.match?("this is my old soccer coach ")
    assert C::RELATION_ATTACHED_BEFORE.match?("our next-door neighbour ")
  end

  def test_the_appositive_and_title_patterns_anchor_at_the_window_start
    # These two are `match` rather than `search` in the reference — the relation
    # must open the window, not merely appear in it. Ruby's `^` would let it open
    # any line of the window instead.
    after = "was there.\n, my cousin, came over"

    refute C::RELATION_ATTACHED_AFTER.match?(after),
           "an appositive on a later line does not follow this name"
    assert loosened(C::RELATION_ATTACHED_AFTER, '\A', "^").match?(after),
           "the line-anchored spelling was expected to match a later line"

    assert C::RELATION_ATTACHED_AFTER.match?(", my cousin, came over")
    assert C::RELATION_ATTACHED_AFTER.match?(", who is my neighbor, came over")
    assert C::TITLE_LEADS_WITH_RELATION.match?("my cousin vinny")
    refute C::TITLE_LEADS_WITH_RELATION.match?("the film\nmy cousin vinny")
  end

  def test_the_sentence_break_anchor_is_inert_here_and_that_is_why_it_stays_strict
    # The odd one out, recorded so nobody "fixes" it in either direction. `^`
    # would report a sentence start after every newline — but the pattern's own
    # `\n+` arm already reports exactly those offsets, and `sentence_starts`
    # returns a Set, so the duplicates collapse and the answer is identical.
    #
    # It is written `\A` anyway: the equivalence is a coincidence of two arms
    # overlapping, not a property anyone should have to re-derive, and it costs a
    # duplicate match per line to rely on.
    loose = loosened(C::SENTENCE_BREAK, '\A', "^")
    ["A line.\nanother line\nthird one.", "one\n\ntwo", "x.\ny? \"Quoted\" z"].each do |text|
      strict_offsets = C.each_match(text, C::SENTENCE_BREAK).map { |m| m.begin(0) + m[0].length }
      loose_offsets = C.each_match(text, loose).map { |m| m.begin(0) + m[0].length }
      assert_equal strict_offsets.to_set, loose_offsets.to_set,
                   "the two anchors were expected to agree as SETS on #{text.inspect}"
      assert_operator loose_offsets.size, :>, strict_offsets.size,
                      "the line anchor was expected to match more often on #{text.inspect}"
    end
  end

  def test_ruby_word_boundaries_already_agree_with_python
    # Recorded because the TypeScript port's reason for spelling these out does
    # NOT apply here, and repeating it would have been a plausible-sounding lie.
    # JavaScript's `\b` is ASCII-only, so it finds a boundary between `n` and `ä`
    # that Python does not. Ruby's `\b` is Unicode-aware and agrees with Python.
    refute_match(/\bcousin\b/, "cousinä",
                 "Ruby's \\b was expected to agree with Python's here")
    refute_match(/\bve/, "naïve")

    # So the explicit lookarounds are belt-and-braces on this side rather than
    # load-bearing — and they stay anyway, because they are the form the shared
    # spec pins and `[\p{L}\p{N}_]` is the same set the gazetteer folds on.
    refute_match(Regexp.new("cousin#{C::NOT_WORD_AFTER}"), "cousinä")
    assert_match(Regexp.new("cousin#{C::NOT_WORD_AFTER}"), "my cousin came over")
    refute_match(Regexp.new("#{C::NOT_WORD_BEFORE}ve"), "naïve")
    assert_match(Regexp.new("#{C::NOT_WORD_BEFORE}ve"), "the ve token")
  end

  def test_the_shorthand_classes_are_ascii_only_which_is_the_real_divergence
    # `\w` is where Ruby actually parts company with Python, and it is why
    # NOT_WORD_BEFORE spells its class out instead of using `\w`.
    refute_match(/\A\w\z/, "ä", "Ruby's \\w was expected to be ASCII-only")
    assert_match(Regexp.new("\\A[#{C::NOT_WORD_BEFORE[4..-2]}]\\z"), "ä")

    # `\d` and `\s` diverge the same way and are deliberately left as-is,
    # matching the TypeScript port, which has the identical narrowing and
    # reproduces every frame. Asserted rather than assumed so the choice is
    # visible: if a future fixture contains a non-ASCII digit, this is the test
    # that says which way the three ports will disagree.
    refute_match(/\A\d\z/, "٣", "Python matches an Arabic-Indic digit here")
    refute_match(/\A\s\z/, " ", "Python matches a non-breaking space here")
  end
end

File: typescript/test/dialect.test.ts
/**
 * Where JavaScript's regex dialect differs from Python's, pinned rather than reasoned about.
 *
 * The counterpart of `ruby/test/dialect_test.rb`. The detector was written against
 * Python `re`; this port transliterated it. Three of JavaScript's defaults disagree
 * with Python's, and each is a silent behaviour change rather than an error:
 *
 * * `\b` is **ASCII-only** in JavaScript and Unicode-aware in Python, so
 *   JavaScript finds a word boundary inside `naïve` and `cousinä` that the
 *   reference never finds.
 * * `\w` and `\d` are ASCII-only here and Unicode-aware in Python.
 * * `$` matches **before a trailing newline** in Python and only at end-of-input
 *   in JavaScript — and gains Ruby's line-anchored meaning the moment the `m` flag
 *   is added.
 *
 * `\s` is the one that does NOT disagree, which is worth a test of its own: it
 * matches a non-breaking space in both, so narrowing it would be a change rather
 * than a correction.
 *
 * Every site is already written the Python way. This file exists because **neither
 * shared spec layer catches it if somebody writes it back.** Measured, not assumed:
 * adding the `m` flag to `ZIP` leaves all 54 conformance frames green and every
 * primitive assertion green, because the fixture corpus is single-line and this
 * rule only diverges across a newline.
 *
 * So each test below asserts the divergence in both directions: that the pattern as
 * written gives the reference answer, AND that the idiomatic-JavaScript spelling
 * gives a different one. The second half is what makes this a test rather than a
 * restatement — an assertion that only pins the current answer would still pass if
 * the difference evaporated, and then it would be guarding nothing.
 *
 * The declared gap this closes is recorded in `conformance/coverage.json`, and
 * `tools/tests/test_coverage_parity.py` fails if its entry outlives it.
 */

import assert from "node:assert/strict";
import { test } from "node:test";

import {
  LOWER_TOKEN,
  NOT_WORD_AFTER,
  NOT_WORD_BEFORE,
  namesSomeoneTheWriterKnows,
} from "../src/candidates.js";
import { ZIP } from "../src/structured.js";

/** The same pattern with a flag added, refusing if it was already there. */
function withFlags(pattern: RegExp, flags: string): RegExp {
  for (const flag of flags) {
    assert.ok(
      !pattern.flags.includes(flag),
      `this pattern already carries the \`${flag}\` flag, so this test is checking nothing`,
    );
  }
  return new RegExp(pattern.source, pattern.flags + flags);
}

/** The same pattern with one piece of its source substituted, refusing a no-op. */
function loosened(pattern: RegExp, from: string, to: string): RegExp {
  const source = pattern.source;
  assert.notEqual(
    source,
    source.replace(from, to),
    `\`${from}\` is no longer in this pattern, so this test is checking nothing`,
  );
  return new RegExp(source.replace(from, to), pattern.flags);
}

function matched(pattern: RegExp, text: string): string[] {
  return [...text.matchAll(pattern)].map((m) => m[0]);
}

// ---------------------------------------------------------------------------
// The word boundary
// ---------------------------------------------------------------------------

test("javascript's word boundary is ASCII-only, which is why the lookarounds are spelled out", () => {
  // The divergence the parity sweep caught before it found agreement. Both
  // directions, because both occur in the source patterns.
  assert.equal(/\bcousin\b/.test("cousinä"), true, "JavaScript's \\b was expected to disagree");
  assert.equal(/\bve/.test("naïve"), true);

  // The spelled-out form agrees with Python, and needs the `u` flag to do it.
  assert.equal(new RegExp(`cousin${NOT_WORD_AFTER}`, "u").test("cousinä"), false);
  assert.equal(new RegExp(`cousin${NOT_WORD_AFTER}`, "u").test("my cousin came over"), true);
  assert.equal(new RegExp(`${NOT_WORD_BEFORE}ve`, "u").test("naïve"), false);
  assert.equal(new RegExp(`${NOT_WORD_BEFORE}ve`, "u").test("the ve token"), true);
});

test("the u flag is not optional decoration on a property escape", () => {
  // The JavaScript-specific trap, and the reason the lookarounds cannot simply be
  // pasted into a pattern that forgot the flag. Without `u`, `\p` is an identity
  // escape: the pattern matches the literal text `p{L}` and NOTHING else, so it
  // silently stops matching letters instead of raising.
  //
  // Built with `new RegExp` rather than written as a literal because tsc REFUSES
  // the literal form (TS1530), which is worth knowing precisely: the compiler
  // already guards the spelling nobody would use here, and the form this port
  // actually uses — a lookaround kept as a string and compiled at module load —
  // is the one it cannot see. So the dynamic form is what gets pinned.
  const noFlag = new RegExp("\\p{L}");
  assert.equal(noFlag.test("Z"), false, "without `u` this must not match a letter");
  assert.equal(noFlag.test("p{L}"), true, "without `u` this is the literal string");
  assert.equal(new RegExp("\\p{L}", "u").test("Z"), true);
  // Which is why every pattern built from these strings carries the flag.
  assert.ok(LOWER_TOKEN.flags.includes("u"), "LOWER_TOKEN must carry the u flag");
  assert.equal(new RegExp(NOT_WORD_BEFORE + "ve").test("naïve"), true, "the unflagged " +
    "lookbehind is inert, which is the silent failure this flag prevents");
});

test("the lowercase route agrees with the reference on an accented word", () => {
  // The end-to-end version of the boundary rule: `ï` is a word character to
  // Python, so the reference emits `na` and stops rather than finding `ve` inside
  // `naïve` — a lowercase token feeding a route whose whole job is deciding what
  // is a name.
  assert.deepEqual(matched(LOWER_TOKEN, "naïve café Renée went home. i did too."), [
    "na",
    "caf",
    "went",
    "home",
    "i",
    "did",
    "too",
  ]);
  // The ASCII control: with no accented character in play the naive spelling
  // agrees, which is what isolates the Unicode difference above.
  const naive = new RegExp("\\b[a-z][a-z'’-]*", "gu");
  assert.deepEqual(
    matched(LOWER_TOKEN, "the cat sat down"),
    matched(naive, "the cat sat down"),
  );
});

test("an accented word tail may not satisfy a relation cue", () => {
  // What the boundary rule buys at the call site. With JavaScript's `\b` the cue
  // `cousin` matches inside `cousinä`, which wrongly overrides a title keep or
  // wrongly refuses corroboration for a public figure's surname.
  assert.equal(
    namesSomeoneTheWriterKnows("naïmy cousin Terrence came over that summer.", 13, 21),
    false,
  );
  assert.equal(
    namesSomeoneTheWriterKnows("Alice Adams, my cousinä came over that summer.", 0, 11),
    false,
  );
  // Genuinely adjacent still attaches, so the guard is not simply switched off.
  assert.equal(
    namesSomeoneTheWriterKnows("My neighbor Alice Adams walked me to the bus stop.", 12, 23),
    true,
  );
});

// ---------------------------------------------------------------------------
// The end-of-input anchor
// ---------------------------------------------------------------------------

test("the multiline flag turns the ZIP anchor into a line anchor and it over-fires", () => {
  // `\s*$` under `m` is satisfied at the end of every line of a hard-wrapped
  // essay, so any 5-digit number a student ends a line on — a locker combination,
  // a population, a year range — masks as a ZIP code.
  const text = "I live at 12345\nMy locker combination is 90210 and I forget it.";

  assert.deepEqual(matched(ZIP, text), [], "a 5-digit number at a line end is not a ZIP");
  assert.deepEqual(
    matched(withFlags(ZIP, "m"), text),
    ["12345"],
    "the line-anchored spelling was expected to over-fire here; if it no longer " +
      "does, this guard has stopped guarding anything",
  );

  // ...and the cases a real ZIP arrives in still match, so the anchor is not
  // simply switched off.
  assert.deepEqual(matched(ZIP, "Send it to 12345."), ["12345"]);
  assert.deepEqual(matched(ZIP, "12345"), ["12345"]);
  assert.deepEqual(matched(ZIP, "Akron 12345 OH"), ["12345"]);
});

test("python's dollar matches before a trailing newline and javascript's does not", () => {
  // The raw divergence, pinned even though `ZIP` is immune to it: `\s*` consumes
  // the newline before `$` is consulted, so the transliteration is correct there
  // by construction rather than by intent. A future pattern using a bare `$`
  // where the reference used one would inherit this silently.
  assert.equal(/12345$/.test("12345\n"), false, "JavaScript's $ is end-of-input");
  // Python's `$` is this, and Python's `\Z` is JavaScript's bare `$`.
  assert.equal(/12345(?=\n?$)/.test("12345\n"), true);
  // Which is why ZIP survives: the whitespace run eats the newline itself.
  assert.deepEqual(matched(ZIP, "Send it to 12345\n"), ["12345"]);
  assert.deepEqual(matched(loosened(ZIP, "\\s*$", "$"), "Send it to 12345\n"), []);
});

// ---------------------------------------------------------------------------
// The shorthand classes
// ---------------------------------------------------------------------------

test("the shorthand classes are ASCII-only, which is the real narrowing", () => {
  // `\w` is where JavaScript actually parts company with Python, and it is why
  // NOT_WORD_BEFORE spells its class out instead of using `\w`.
  assert.equal(/^\w$/u.test("ä"), false, "JavaScript's \\w was expected to be ASCII-only");
  assert.equal(new RegExp(`^[${"\\p{L}\\p{N}_"}]$`, "u").test("ä"), true);

  // `\d` diverges the same way and is deliberately left as-is, matching the Ruby
  // port, which has the identical narrowing and reproduces every frame. Asserted
  // rather than assumed so the choice is visible: if a future fixture contains a
  // non-ASCII digit, this is the test that says which way the three ports will
  // disagree.
  assert.equal(/^\d$/u.test("٣"), false, "Python matches an Arabic-Indic digit here");
});

test("the whitespace class is the one that already agrees, so narrowing it would be a change", () => {
  // Recorded because the reasoning above does NOT extend to `\s`, and extending it
  // would have been a plausible-sounding lie. Both languages match a non-breaking
  // space, so the patterns that use `\s` need no correction.
  assert.equal(/^\s$/u.test(" "), true, "JavaScript's \\s matches a non-breaking space");
  assert.equal(/^\s$/u.test(" "), true, "...and an em space");
});
